### PR TITLE
Proposal to clean up ArrayData and function dispatch

### PR DESCRIPTION
**Pull request by gatesn**

### CODE REVIEW

File: vortex-array/src/array2.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    fn is_valid(&self, index: usize) -> VortexResult<bool> {
+        (self.encoding.is_valid)(self, index)
+    }
+}
+
+pub struct EncodingVTable {
+    id: &'static str,
+    into_canonical: &'static dyn Fn(ArrayData) -> VortexResult<Canonical>,
+    is_valid: &'static dyn Fn(&ArrayData, usize) -> VortexResult<bool>,
```
Comment (a10y) [inline vortex-array/src/array2.rs]: could we just use function pointers for these, e.g.

```
into_canonical: fn(ArrayData) -> VortexResult<Canonical>
```

then it works for both functions and non-capturing closures?

Comment (robert3005) [reply vortex-array/src/array2.rs]: `&'static dyn Fn` is a very particular type in Rust. I think what @a10y is the thing you want

File: vortex-array/src/array2.rs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+    fn is_valid(&self, index: usize) -> VortexResult<bool> {
+        (self.encoding.is_valid)(self, index)
+    }
+}
+
+pub struct EncodingVTable {
+    id: &'static str,
+    into_canonical: fn(data: ArrayData) -> VortexResult<Canonical>,
+    is_valid: fn(data: &ArrayData, idx: usize) -> VortexResult<bool>,
```
Comment (a10y) [inline vortex-array/src/array2.rs:70]: why use the type-erased thing? this is the EncodingVTable, but every implementation needs to check the encoding of the thing it's given?

Comment (gatesn) [reply vortex-array/src/array2.rs:70]: The vtable _must_ be type erased. You can't `impl Array for ArrayData` otherwise

**Pull request closed without merging**